tree/dataframe/src/RNTupleDS.cxx

```diff
@@ -69,27 +133,120 @@ class RNTupleColumnReader final : public ROOT::Detail::RDF::RColumnReaderBase {
       return fValue.GetRawPtr();
    }
 };
+
 } // namespace Detail
 
-void RNTupleDS::AddFields(const RNTupleDescriptor &desc, DescriptorId_t parentId)
+
+RNTupleDS::~RNTupleDS() = default;
+
+
+void RNTupleDS::AddField(
+   const RNTupleDescriptor &desc, std::string_view colName, DescriptorId_t fieldId,
+   std::vector<DescriptorId_t> skeinIDs)
 {
-   for (const auto& f : desc.GetFieldRange(parentId)) {
-      fColumnNames.emplace_back(desc.GetQualifiedFieldName(f.GetId()));
-      fColumnTypes.emplace_back(f.GetTypeName());
-      if (f.GetStructure() == ENTupleStructure::kRecord)
-         AddFields(desc, f.GetId());
+   // As an example for the mapping of RNTuple fields to RDF columns, let's consider an RNTuple
+   // using the following types and with a top-level field named "event" of type Event:
+   //
+   // struct Event {
+   //    int id;
+   //    std::vector<Track> tracks;
+   // };
+   // struct Track {
+   //    std::vector<Hit> hits;
+   // };
+   // struct Hit {
+   //    float x;
+   //    float y;
+   // };
+   //
+   // AddField() will be called from the constructor with the RNTuple root field (ENTupleStructure::kRecord).
+   // From there, we recurse into the "event" sub field (also ENTupleStructure::kRecord) and further down the
+   // tree of sub fields and expose the following RDF columns:
+   // TODO(jblomer): Collections should be exposed as RVec<T> instead of std::vector<T>
+   //
+   // "event"                             [Event]
+   // "event.id"                          [int]
+   // "event.tracks"                      [std::vector<Track>]
+   // "__rdf_sizeof_event.tracks"         [unsigned int]
+   // "event.tracks.hits"                 [std::vector<std::vector<Hit>>]
+   // "__rdf_sizeof_event.tracks.hits"    [std::vector<unsigned int>]
+   // "event.tracks.hits.x"               [std::vector<std::vector<float>>]
+   // "__rdf_sizeof_event.tracks.hits.x"  [std::vector<unsigned int>]
+   // "event.tracks.hits.y"               [std::vector<std::vector<float>>]
+   // "__rdf_sizeof_event.tracks.hits.y"  [std::vector<unsigned int>]
```

Comment:
The problem is: which size? The sum of hit sizes over all tracks?

Reply:
My mental model is "`#var` is a performant shorthand for `var.size()`", so `#event.tracks.hits` would actually be equivalent to `#event.tracks` (because the `vector<Track>` and the `vector<vector<float>>` return the same number when you call `.size()` on them).

Your version probably makes more sense :smile: but I don't know how to explain it in plain english to users.

Reply:
will have to be documented very well, but it makes sense as it is